File: src/main/java/org/ym6/TuneSet.java
package org.ym6;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import org.jspecify.annotations.Nullable;
import java.util.ArrayList;
import java.util.List;
import org.ymx.Tools;

/**
 * What a set of YM dumps says about itself, for the SNDH tags: each tune's
 * name, a composer they all share, and a title made of the lot.
 *
 * <p>The dumps are wildly inconsistent about their own metadata, so a name
 * that says nothing - empty, "unknown", "untitled" - gives way to the file
 * stem, which at least came from someone who knew what the tune was. A
 * composer is only claimed when every tune in the set agrees on one; a set
 * assembled from different musicians has no single COMM to declare.
 */
public record TuneSet(List<String> names, @Nullable String composer, String title) {

    public static TuneSet of(List<Path> tunes) {
        List<String> names = new ArrayList<>();
        @Nullable String composer = null;
        boolean agree = true;
        for (Path tune : tunes) {
            Ym6Reader.Song song = read(tune);
            String name = song.name().strip();
            if (saysNothing(name)) {
                name = stem(tune);
            }
            names.add(name);
            String author = song.author().strip();
            if (composer == null && !author.isEmpty()) {
                composer = author;
            } else if (!author.equals(composer == null ? "" : composer)) {
                agree = false;
            }
        }
        return new TuneSet(names, agree ? composer : null, String.join(" / ", names));
    }

    /** The rate every tune must share: one SNDH declares one. */
    public static int playerHz(Path tune) {
        return read(tune).playerHz();
    }

    private static Ym6Reader.Song read(Path tune) {
        try {
            return Ym6Reader.read(Files.readAllBytes(tune));
        } catch (IOException | Ym6Reader.FormatException e) {
            throw Tools.fail(tune + ": " + e.getMessage());
        }
    }

    static String stem(Path tune) {
        return tune.getFileName().toString().replaceAll("(?i)\\.ym$", "");
    }

    private static boolean saysNothing(String name) {
        String lower = name.toLowerCase(java.util.Locale.ROOT);
        return lower.isEmpty() || lower.equals("unknown") || lower.equals("untitled")
                || lower.equals("<unknown>");
    }
}

File: go/internal/pack/tuneset.go
package pack

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/odipar/ymx/internal/ym"
)

// TuneSet is what a set of dumps calls itself: each tune's name, a composer
// they all share, and a title made of the lot. A name that says nothing gives
// way to the file stem; a composer is only claimed when every tune agrees.
type TuneSet struct {
	Names    []string
	Composer string
	Title    string
}

// SetOf reads the dumps and says what the set calls itself.
func SetOf(command string, tunes []string) (TuneSet, error) {
	names := make([]string, 0, len(tunes))
	composer := ""
	claimed := false
	agree := true
	for _, tune := range tunes {
		song, err := readSong(command, tune)
		if err != nil {
			return TuneSet{}, err
		}
		name := strings.TrimSpace(song.Name)
		if saysNothing(name) {
			name = Stem(tune)
		}
		names = append(names, name)
		author := strings.TrimSpace(song.Author)
		if !claimed && author != "" {
			composer, claimed = author, true
		} else if author != composer {
			agree = false
		}
	}
	set := TuneSet{Names: names, Title: strings.Join(names, " / ")}
	if agree {
		set.Composer = composer
	}
	return set, nil
}

// PlayerHz is the rate every tune must share: one SNDH declares one.
func PlayerHz(command, tune string) (int, error) {
	song, err := readSong(command, tune)
	if err != nil {
		return 0, err
	}
	return song.PlayerHz, nil
}

// Stem is a dump's file name without its extension.
func Stem(tune string) string {
	name := filepath.Base(tune)
	if strings.EqualFold(filepath.Ext(name), ".ym") {
		return name[:len(name)-3]
	}
	return name
}

// readSong reads one dump. A file that does not open names the command,
// because the fault is the command line; a file that opens and is not a
// dump names the file, because the fault is in it. The other two trees
// split it the same way, and nothing of the system call reaches either
// message.
func readSong(command, tune string) (*ym.Song, error) {
	data, err := os.ReadFile(tune)
	if err != nil {
		return nil, fmt.Errorf("%s: cannot read %s", command, tune)
	}
	song, err := ym.Read(data)
	if err != nil {
		return nil, fmt.Errorf("%s: %w", tune, err)
	}
	return song, nil
}

func saysNothing(name string) bool {
	lower := strings.ToLower(name)
	return lower == "" || lower == "unknown" || lower == "untitled" ||
		lower == "<unknown>"
}
